pkg/querier/http.go

```diff
@@ -126,3 +126,50 @@ func (q *Querier) LabelHandler(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 }
+
+// TailHandler is a http.HandlerFunc for handling tail queries.
+func (q *Querier) TailHandler(w http.ResponseWriter, r *http.Request) {
+	upgrader := websocket.Upgrader{
+		CheckOrigin: func(r *http.Request) bool { return true },
+	}
+
+	conn, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		level.Error(util.Logger).Log("Error in upgrading websocket", fmt.Sprintf("%v", err))
+		return
+	}
+
+	defer func() {
+		err := conn.Close()
+		level.Error(util.Logger).Log("Error closing websocket", fmt.Sprintf("%v", err))
+	}()
+
+	queryRequest, err := httpRequestToQueryRequest(r)
+	if err != nil {
+		if err := conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseUnsupportedData, err.Error())); err != nil {
+			level.Error(util.Logger).Log("Error writing close message to websocket", fmt.Sprintf("%v", err))
+		}
+		return
+	}
+
+	itr := q.tailQuery(r.Context(), &queryRequest)
+	stream := logproto.Stream{}
+
+	for itr.Next() {
+		stream.Entries = []logproto.Entry{itr.Entry()}
+		stream.Labels = itr.Labels()
+
+		err := conn.WriteJSON(stream)
+		if err != nil {
+			level.Error(util.Logger).Log("Error writing to websocket", fmt.Sprintf("%v", err))
+			break
+		}
+	}
+
+	if err := itr.Error(); err != nil {
+		level.Error(util.Logger).Log("Error from iterator", fmt.Sprintf("%v", err))
+		if err := conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseInternalServerErr, err.Error())); err != nil {
+			level.Error(util.Logger).Log("Error writing close message to websocket", fmt.Sprintf("%v", err))
```

Comment:
For this kinda thing (just logging the error if it fails) we have a helper function - `helpers.LogError`.